decoder.go

```diff
@@ -20,18 +60,56 @@ func frameLineSize(frame *C.AVFrame) *C.int {
 	return (*C.int)(unsafe.Pointer(&frame.linesize[0]))
 }
 
-// h264Decoder is a wrapper around ffmpeg's H264 decoder.
-type h264Decoder struct {
-	codecCtx    *C.AVCodecContext
-	srcFrame    *C.AVFrame
-	swsCtx      *C.struct_SwsContext
-	dstFrame    *C.AVFrame
-	dstFramePtr []uint8
+// getStreamInfo opens a stream URL and retrieves the video codec.
+func getStreamInfo(url string) (videoCodec, error) {
+	cUrl := C.CString(url)
+	defer C.free(unsafe.Pointer(cUrl))
+
+	var avFormatCtx *C.AVFormatContext = nil
+	ret := C.avformat_open_input(&avFormatCtx, cUrl, nil, nil)
+	if ret < 0 {
+		return Unknown, fmt.Errorf("avformat_open_input() failed: %s", avError(ret))
+	}
+	defer C.avformat_close_input(&avFormatCtx)
+
+	ret = C.avformat_find_stream_info(avFormatCtx, nil)
+	if ret < 0 {
+		return Unknown, fmt.Errorf("avformat_find_stream_info() failed: %s", avError(ret))
+	}
+
+	cCodec := C.get_video_codec(avFormatCtx)
+	codec := convertCodec(cCodec)
+
+	if codec == Unknown {
+		return Unknown, errors.New("no supported codec found")
+	}
+	return codec, nil
 }
 
-// newH264Decoder allocates a new h264Decoder.
-func newH264Decoder() (*h264Decoder, error) {
-	codec := C.avcodec_find_decoder(C.AV_CODEC_ID_H264)
+// convertCodec converts a C int to a Go videoCodec.
+func convertCodec(cCodec C.int) videoCodec {
+	switch cCodec {
+	case C.AV_CODEC_ID_H264:
+		return H264
+	case C.AV_CODEC_ID_HEVC:
+		return H265
+	default:
+		return Unknown
+	}
+}
+
+// avError converts an AV error code to a AV error message string.
+func avError(avErr C.int) string {
+	var errbuf [1024]C.char
+	if C.av_strerror(avErr, &errbuf[0], 1024) < 0 {
```

Comment:
Any update on what the behavior is if the buffer is not big enough to write the error?

Reply:
https://ffmpeg.org/doxygen/trunk/group__lavu__error.html#ga5792b4a2d18d7d9cb0efbcfc335dce24

Here is the `av_strerror` fn. It uses `av_strlcpy` which has safety features ontop of `strcpy`. So if msg is greater than 1023 (1024 - 1 for null term), it will be truncated.

Reply:
On further review, it looks like there is a hardcoded macro for max length of 64.
https://ffmpeg.org/doxygen/trunk/group__lavu__error.html#ga326c3f741a7adb148cdeaad68eea713d

Reply:
ok, updated so that it uses the `AV_ERROR_MAX_STRING_SIZE` macro directly instead of hardcoding 1024.